File: C/src/makemove.c
// makemove.cpp

#include "makemove.h"
#include "Board.h"
#include "bitboard.h"
#include "movegen.h"
#include "attack.h"

// Functions based on VICE makemove.c by Richard Allbert
// Incrementally updating Board class move by move for better efficiency

// castling rights update constants
const int castling_rights[64] = {
     7, 15, 15, 15,  3, 15, 15, 11,
    15, 15, 15, 15, 15, 15, 15, 15,
    15, 15, 15, 15, 15, 15, 15, 15,
    15, 15, 15, 15, 15, 15, 15, 15,
    15, 15, 15, 15, 15, 15, 15, 15,
    15, 15, 15, 15, 15, 15, 15, 15,
    15, 15, 15, 15, 15, 15, 15, 15,
    13, 15, 15, 15, 12, 15, 15, 14
};

/*
	Piece manipulation
*/

static inline void clear_piece(Board *pos, const int sq) {
	int pce = pos->pieces[sq];
	int col = piece_col[pce];

	pos->pieces[sq] = EMPTY;
	pos->piece_num[pce]--;

	CLR_BIT(pos->bitboards[pce], sq);
	CLR_BIT(pos->occupancies[col], sq);
	CLR_BIT(pos->occupancies[BOTH], sq);
}

static void add_piece(Board *pos, const int sq, const int pce) {
	int col = piece_col[pce];

	pos->pieces[sq] = pce;
	pos->piece_num[pce]++;

	SET_BIT(pos->bitboards[pce], sq);
	SET_BIT(pos->occupancies[col], sq);
	SET_BIT(pos->occupancies[BOTH], sq);
}

static void move_piece(Board *pos, const int from, const int to) {
	int pce = pos->pieces[from];
	int col = piece_col[pce];

	pos->pieces[from] = EMPTY;
	CLR_BIT(pos->bitboards[pce], from);
	CLR_BIT(pos->occupancies[col], from);
	CLR_BIT(pos->occupancies[BOTH], from);

	pos->pieces[to] = pce;
	SET_BIT(pos->bitboards[pce], to);
	SET_BIT(pos->occupancies[col], to);
	SET_BIT(pos->occupancies[BOTH], to);
}

/*
	Move manipulation
*/

void take_move(Board *pos) {
	pos->ply--;
	pos->his_ply--;
	
	UndoBox box = pos->move_history[pos->his_ply];
	int move = box.move;
	int from = get_move_source(move);
	int to = get_move_target(move);

	pos->castle_perms = box.castle_perms;
	pos->fifty_move = box.fifty_move;
	pos->enpas = box.enpas;

	pos->side ^= 1;

	if (get_move_enpassant(move)) {
		// Recover the pawn that was enpassanted
		if (pos->side == WHITE) {
			add_piece(pos, to + 8, bP);
		}
		else {
			add_piece(pos, to - 8, wP);
		}
	}
	else if (get_move_castling(move)) {
		// Move the castling rook back
		switch (to) {
			case c1: move_piece(pos, d1, a1); break;
			case c8: move_piece(pos, d8, a8); break;
			case g1: move_piece(pos, f1, h1); break;
			case g8: move_piece(pos, f8, h8); break;
		}
	}

	move_piece(pos, to, from);

	if (piece_type[pos->pieces[from]] == KING) {
		pos->king_sq[pos->side] = from;
	}

	// Undo captures
	int captured = get_move_captured(move);
	if (captured != EMPTY && !get_move_enpassant(move)) {
		add_piece(pos, to, captured);
	}

	// Undo promotion
	if (get_move_promoted(move) != EMPTY) {
		clear_piece(pos, from);
		add_piece(pos, from, (piece_col[get_move_promoted(move)] == WHITE) ? wP : bP);
	}
}

// Makes a move on the board
// Returns true if legal, and false if illegal
bool make_move(Board* pos, int move) {
	int from = get_move_source(move);
	int to = get_move_target(move);
	int side = pos->side;

	UndoBox box = { 0, 0, 0, 0, 0 };

	if (get_move_enpassant(move)) {
		// Clear the captured pawn
		if (side == WHITE) {
			clear_piece(pos, to + 8);
		}
		else {
			clear_piece(pos, to - 8);
		}
	}
	else if (get_move_castling(move)) {
		// Move the corresponding rook
		switch (to) {
		case c1:
			move_piece(pos, a1, d1);
			break;
		case c8:
			move_piece(pos, a8, d8);
			break;
		case g1:
			move_piece(pos, h1, f1);
			break;
		case g8:
			move_piece(pos, h8, f8);
			break;
		}
	}

	box.move = move;
	box.fifty_move = pos->fifty_move;
	box.enpas = pos->enpas;
	box.castle_perms = pos->castle_perms;
	pos->move_history[pos->his_ply] = box;

	pos->castle_perms &= castling_rights[from];
	pos->castle_perms &= castling_rights[to];
	pos->enpas = NO_SQ;

	pos->fifty_move++;

	// Handle captures
	int captured = get_move_captured(move);
	if (captured) {
		clear_piece(pos, to);
		pos->fifty_move = 0; // A capture is played - reset 50-move counter
	}

	pos->ply++;
	pos->his_ply++;

	if (piece_type[pos->pieces[from]] == PAWN) {
		pos->fifty_move = 0; // A pawn is moved - reset 50-move counter
		// Check if it's double advance
		if (get_move_double(move)) {
			if (side == WHITE) {
				pos->enpas = from - 8;
			}
			else {
				pos->enpas = from + 8;
			}
		}
	}

	move_piece(pos, from, to);

	// Handle promotions
	int prPce = get_move_promoted(move);
	if (prPce != EMPTY) {
		// Replace the pawn with the promoted piece
		clear_piece(pos, to);
		add_piece(pos, to, prPce);
	}

	// Detect king move
	if (piece_type[pos->pieces[to]] == KING) {
		pos->king_sq[pos->side] = to;
	}

	// Switch sides
	pos->side ^= 1;

	// It is an illegal move if we reveal a check to our king
	// side: Our side (the mover)
	// pos->get_side(): Opponent's side (switched after making the move)
	if (is_square_attacked(pos, pos->king_sq[side], pos->side)) {
		take_move(pos);
		return FALSE; // Illegal move
	}

	return TRUE;
}

File: Python/src/makemove.py
# makemove.cpp

from makemove_helper import *
from Board import Board, UndoBox
from datatypes import Piece, PType, Squares, Colour, piece_col, piece_type
from Move import Move
from attack import is_square_attacked
from bitboard import *

# Functions based on VICE makemove.c by Richard Allbert
# Incrementally updating Board class move by move for better efficiency

"""
	Piece manipulation
"""

def clear_piece(pos: Board, sq: Squares) -> Board:
    pce = pos.pieces[sq.value]
    col = piece_col[pce]

    pos.pieces[sq.value] = Piece.EMPTY
    pos.piece_num[pce.value] -= 1

    pos.bitboards[pce.value] = clear_bit(pos.bitboards[pce.value], sq)
    pos.occupancies[col.value] = clear_bit(pos.occupancies[col.value], sq)
    pos.occupancies[Colour.BOTH.value] = clear_bit(pos.occupancies[Colour.BOTH.value], sq)
    return pos

def add_piece(pos: Board, sq: Squares, pce: Piece) -> Board:
	col = piece_col[pce]

	pos.pieces[sq.value] = pce
	pos.piece_num[pce.value] += 1

	pos.bitboards[pce.value] = set_bit(pos.bitboards[pce.value], sq)
	pos.occupancies[col.value] = set_bit(pos.occupancies[col.value], sq)
	pos.occupancies[Colour.BOTH.value] = set_bit(pos.occupancies[Colour.BOTH.value], sq)
	return pos

# "from" is a Python keyword, so we renamed it here
def move_piece(pos: Board, from_sq: Squares, to_sq: Squares) -> Board:
    pce = pos.pieces[from_sq.value]
    col = piece_col[pce]

    pos.pieces[from_sq.value] = Piece.EMPTY
    pos.bitboards[pce.value] = clear_bit(pos.bitboards[pce.value], from_sq)
    pos.occupancies[col.value] = clear_bit(pos.occupancies[col.value], from_sq)
    pos.occupancies[Colour.BOTH.value] = clear_bit(pos.occupancies[Colour.BOTH.value], from_sq)

    pos.pieces[to_sq.value] = pce
    pos.bitboards[pce.value] = set_bit(pos.bitboards[pce.value], to_sq)
    pos.occupancies[col.value] = set_bit(pos.occupancies[col.value], to_sq)
    pos.occupancies[Colour.BOTH.value] = set_bit(pos.occupancies[Colour.BOTH.value], to_sq)
    return pos

"""
	Move manipulation
"""

def take_move(pos: Board) -> Board:
    pos.ply -= 1
    pos.his_ply -= 1

    box = pos.move_history[pos.his_ply]
    move = box.move
    from_square = move.source
    to_square = move.target

    pos.castle_perms = box.castle_perms
    pos.fifty_move = box.fifty_move
    pos.enpas = box.enpas

    pos.side = Colour(pos.side.value ^ 1)

    if move.enpassant:
        # Recover the pawn that was enpassanted
        if pos.side == Colour.WHITE:
            pos = add_piece(pos, Squares(to_square.value + 8), Piece.bP)
        else:
            pos = add_piece(pos, Squares(to_square.value - 8), Piece.wP)
            
    elif move.castling:
        # Move the castling rook back
        if to_square == Squares.c1:
            pos = move_piece(pos, Squares.d1, Squares.a1)
        elif to_square == Squares.c8:
            pos = move_piece(pos, Squares.d8, Squares.a8)
        elif to_square == Squares.g1:
            pos = move_piece(pos, Squares.f1, Squares.h1)
        elif to_square == Squares.g8:
            pos = move_piece(pos, Squares.f8, Squares.h8)

    pos = move_piece(pos, to_square, from_square)

    if piece_type[pos.pieces[from_square.value]] == PType.KING:
        pos.king_sq[pos.side.value] = from_square

    # Undo captures
    captured = move.captured
    if captured != Piece.EMPTY and not move.enpassant:
        pos = add_piece(pos, to_square, captured)

    # Undo promotion
    if move.promoted != Piece.EMPTY:
        pos = clear_piece(pos, from_square)
        pos = add_piece(pos, from_square, Piece.wP if piece_col[move.promoted] == Colour.WHITE else Piece.bP)

    return pos

def make_move(pos: Board, move: Move) -> tuple[bool, Board]:
    from_square = move.source
    to_square = move.target
    side = pos.side

    box = UndoBox()

    if move.enpassant:
        # Clear the captured pawn
        if side == Colour.WHITE:
            pos = clear_piece(pos, Squares(to_square.value + 8))
        else:
            pos = clear_piece(pos, Squares(to_square.value - 8))
            
    elif move.castling:
        # Move the corresponding rook
        if to_square == Squares.c1:
            pos = move_piece(pos, Squares.a1, Squares.d1)
        elif to_square == Squares.c8:
            pos = move_piece(pos, Squares.a8, Squares.d8)
        elif to_square == Squares.g1:
            pos = move_piece(pos, Squares.h1, Squares.f1)
        elif to_square == Squares.g8:
            pos = move_piece(pos, Squares.h8, Squares.f8)

    box.move = move
    box.fifty_move = pos.fifty_move
    box.enpas = pos.enpas
    box.castle_perms = pos.castle_perms
    pos.move_history[pos.his_ply] = box

    pos.castle_perms &= castling_rights[from_square.value]
    pos.castle_perms &= castling_rights[to_square.value]
    pos.enpas = Squares.NO_SQ

    pos.fifty_move += 1

    # Handle captures
    captured = move.captured
    if captured:
        pos = clear_piece(pos, to_square)
        pos.fifty_move = 0  # A capture is played - reset 50-move counter

    pos.ply += 1
    pos.his_ply += 1

    if piece_type[pos.pieces[from_square.value]] == PType.PAWN:
        pos.fifty_move = 0  # A pawn is moved - reset 50-move counter
        # Check if it's a double advance
        if move.double_push:
            pos.enpas = Squares(from_square.value - 8) if side == Colour.WHITE else Squares(from_square.value + 8)

    pos = move_piece(pos, from_square, to_square)

    # Handle promotions
    prPce = move.promoted
    if prPce != Piece.EMPTY:
        # Replace the pawn with the promoted piece
        pos = clear_piece(pos, to_square)
        pos = add_piece(pos, to_square, prPce)

    # Detect king move
    if piece_type[pos.pieces[to_square.value]] == PType.KING:
        pos.king_sq[pos.side.value] = to_square

    # Switch sides
    pos.side = Colour(pos.side.value ^ 1)

    # It is an illegal move if we reveal a check to our king
    if is_square_attacked(pos, pos.king_sq[side.value], pos.side):
        pos = take_move(pos)
        return False, pos  # Illegal move

    return True, pos